torch/csrc/generic/Tensor.cpp

```diff
@@ -208,6 +209,7 @@ THTensor* THPTensor_(fromNumpy)(PyObject *numpy_array) {
       switch (PyArray_TYPE(array)) {
         case NPY_DOUBLE: COPY_FROM_ARRAY(double,  array, storage, storage_size); break;
         case NPY_FLOAT:  COPY_FROM_ARRAY(float,   array, storage, storage_size); break;
+        case NPY_HALF:   COPY_FROM_ARRAY(float,   array, storage, storage_size); break;
```

Comment:
I don't think that will work. It will cast the data pointer to float and use this to do assignments. What I meant is that you would change these `CUDA_COPY_FROM_ARRAY` macros to be much more similar to the CPU-side ones (i.e. just do a cudaMemcpy). Do you have experience with CUDA programming and know how to do this? If not I can fix it on top of your commits.